### PR TITLE
Native astype

### PR DESCRIPTION
**Pull request by Rubtsowa**

### CODE REVIEW

File: dpnp/backend/kernels/dpnp_krnl_common.cpp [hunk 1]
```diff
@@ -35,6 +35,42 @@
 namespace mkl_blas = oneapi::mkl::blas;
 namespace mkl_lapack = oneapi::mkl::lapack;
 
+template <typename _DataType, typename _TypeNew, typename _ResultType>
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_common.cpp:38]: what is `_TypeNew` for?

File: dpnp/backend/kernels/dpnp_krnl_common.cpp [hunk 2]
```diff
@@ -35,6 +35,42 @@
...
+class dpnp_astype_c_kernel;
+
+template <typename _DataType, typename _TypeNew, typename _ResultType>
+void dpnp_astype_c(void* array1_in, void* result1, size_t size)
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_common.cpp:42]: remember this interface only for `unsafe` casting kind.

File: dpnp/backend/kernels/dpnp_krnl_common.cpp [hunk 3]
```diff
@@ -35,6 +35,42 @@
...
+{
+    cl::sycl::event event;
+
+    _DataType* array_in = reinterpret_cast<_DataType*>(array1_in);
+    _ResultType* result = reinterpret_cast<_ResultType*>(result1);
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_common.cpp:47]:
```suggestion
void dpnp_astype_c(const void* array1_in, void* result1, const size_t size)
{
    cl::sycl::event event;

    const _DataType* array_in = reinterpret_cast<const _DataType*>(array1_in);
    _ResultType* result = reinterpret_cast<_ResultType*>(result1);
```
Try not to ignore suggestions from previous PRs.

File: dpnp/dparray.pyx [hunk 4]
```diff
@@ -788,19 +790,21 @@ cdef class dparray:
...
+                pass
+            elif copy is not True:
+                pass
+            elif order is not 'K':
+                pass
+            else:
+                return dpnp_astype(self, dtype)
 
-        return dpnp_astype(self, dtype)
+        return call_origin(numpy.ndarray.astype, self, dtype, order, casting, subok, copy)
```
Comment (samir-nasibli) [inline dpnp/dparray.pyx:807]: I think doing fallback on `numpy.ndarray` here is bad idea.
It is better just throw error.

File: tests_external/skipped_tests_numpy.tbl [hunk 5]
```diff
@@ -2881,6 +2881,7 @@ tests/test_umath.py::test_rint_big_int
 tests/test_umath.py::TestRoundingFunctions::test_object_direct
 tests/test_umath.py::TestRoundingFunctions::test_object_indirect
 tests/test_umath.py::test_signaling_nan_exceptions
+tests/test_umath.py::TestSign::test_sign_dtype_nan_object
```
Comment (samir-nasibli) [inline tests_external/skipped_tests_numpy.tbl:2884]: what is the reason?
Provide info on description, please.

File: tests/test_dparray.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import pytest
+
+
+@pytest.mark.parametrize("res_dtype",
+                         [numpy.float64, numpy.float32, numpy.int64, numpy.int32, numpy.bool, numpy.bool_, numpy.complex],
+                         ids=['float64', 'float32', 'int64', 'int32', 'bool', 'bool_', 'complex'])
+@pytest.mark.parametrize("arr_dtype",
+                         [numpy.float64, numpy.float32, numpy.int64, numpy.int32, numpy.bool, numpy.bool_, numpy.complex],
+                         ids=['float64', 'float32', 'int64', 'int32', 'bool', 'bool_', 'complex'])
+@pytest.mark.parametrize("arr",
```
Comment (samir-nasibli) [inline tests/test_dparray.py:12]: what if array is empty?
Corner cases.

File: tests/skipped_tests_gpu.tbl [hunk 7]
```diff
@@ -1,3 +1,17 @@
...
+tests/test_dparray.py::test_astype[[-2, -1, 0, 1, 2]-complex-bool]
+tests/test_dparray.py::test_astype[[-2, -1, 0, 1, 2]-complex-bool_]
+tests/test_dparray.py::test_astype[[-2, -1, 0, 1, 2]-complex-complex]
+tests/test_dparray.py::test_astype[[[-2, -1], [1, 2]]-complex-float64]
+tests/test_dparray.py::test_astype[[[-2, -1], [1, 2]]-complex-float32]
+tests/test_dparray.py::test_astype[[[-2, -1], [1, 2]]-complex-int64]
+tests/test_dparray.py::test_astype[[[-2, -1], [1, 2]]-complex-int32]
+tests/test_dparray.py::test_astype[[[-2, -1], [1, 2]]-complex-bool]
+tests/test_dparray.py::test_astype[[[-2, -1], [1, 2]]-complex-bool_]
+tests/test_dparray.py::test_astype[[[-2, -1], [1, 2]]-complex-complex]
```
Comment (samir-nasibli) [inline tests/skipped_tests_gpu.tbl:14]: please, provide some info why are you skipping all this tests?

File: tests/skipped_tests.tbl [hunk 8]
```diff
@@ -1,3 +1,17 @@
...
+tests/test_dparray.py::test_astype[[-2, -1, 0, 1, 2]-complex-bool]
+tests/test_dparray.py::test_astype[[-2, -1, 0, 1, 2]-complex-bool_]
+tests/test_dparray.py::test_astype[[-2, -1, 0, 1, 2]-complex-complex]
+tests/test_dparray.py::test_astype[[[-2, -1], [1, 2]]-complex-float64]
+tests/test_dparray.py::test_astype[[[-2, -1], [1, 2]]-complex-float32]
+tests/test_dparray.py::test_astype[[[-2, -1], [1, 2]]-complex-int64]
+tests/test_dparray.py::test_astype[[[-2, -1], [1, 2]]-complex-int32]
+tests/test_dparray.py::test_astype[[[-2, -1], [1, 2]]-complex-bool]
+tests/test_dparray.py::test_astype[[[-2, -1], [1, 2]]-complex-bool_]
+tests/test_dparray.py::test_astype[[[-2, -1], [1, 2]]-complex-complex]
```
Comment (samir-nasibli) [inline tests/skipped_tests.tbl:14]: what is the reason?

**Review by samir-nasibli — CHANGES_REQUESTED** [review]
Provide also info about tests, that you are disabling.

File: tests/skipped_tests.tbl (hunk 8 above)
Comment (Rubtsowa) [reply tests/skipped_tests.tbl:14]: because 

error: assigning to 'bool' from incompatible type 'const std::complex<double>'
error: assigning to 'int' from incompatible type 'const std::complex<double>'
error: assigning to 'long' from incompatible type 'const std::complex<double>'
error: assigning to 'float' from incompatible type 'const std::complex<double>'
error: assigning to 'double' from incompatible type 'const std::complex<double>'

**Comment by samir-nasibli** [comment]
So seems that, you are disabling even more tests for changes proposed by you.
You need to investigate them.

File: tests/skipped_tests.tbl (hunk 8 above)
Comment (samir-nasibli) [reply tests/skipped_tests.tbl:14]: Seems like they have the same issue. Investigation and fix is needed.

**Comment by Rubtsowa** [comment]
@samir-nasibli  you sure that I am disabling even more tests?
I added 196 tests, 20 of them skipped(not converted many types in complex in c++ ). And I uskip 1 test and 1 skip in existing tests.

File: tests/skipped_tests_gpu.tbl [hunk 9]
```diff
@@ -1,3 +1,24 @@
+tests/test_dparray.py::test_astype[[-2, -1, 0, 1, 2]-complex-float64]
+tests/test_dparray.py::test_astype[[-2, -1, 0, 1, 2]-complex-float32]
+tests/test_dparray.py::test_astype[[-2, -1, 0, 1, 2]-complex-int64]
+tests/test_dparray.py::test_astype[[-2, -1, 0, 1, 2]-complex-int32]
+tests/test_dparray.py::test_astype[[-2, -1, 0, 1, 2]-complex-bool]
+tests/test_dparray.py::test_astype[[-2, -1, 0, 1, 2]-complex-bool_]
+tests/test_dparray.py::test_astype[[-2, -1, 0, 1, 2]-complex-complex]
```
Comment (samir-nasibli) [inline tests/skipped_tests_gpu.tbl:7]: Complex to complex doesn't works?

File: tests_external/skipped_tests_numpy.tbl (hunk 5 above)
Comment (Rubtsowa) [reply tests_external/skipped_tests_numpy.tbl:2884]: This test failed with error:
DPNP error: in function dpnp_dtype_to_DPNPFuncType() type '<class 'object'>' is not supported

Comment (samir-nasibli) [reply tests_external/skipped_tests_numpy.tbl:2884]: `object` is a (global) variable. By default it is bound to a built-in class which is the root of the type hierarchy.

`dpnp_dtype_to_DPNPFuncType` can not convert `object` to `dpnpc` function types (DPNP_FT_DOUBLE, DPNP_FT_FLOAT, DPNP_FT_LONG, ...). 
See this internal func impl [here (https://github.com/IntelPython/dpnp/blob/master/dpnp/dpnp_algo/dpnp_algo.pyx#L256:L271)

You need fallback on `numpy` for case when dtype is object.

Comment (samir-nasibli) [reply tests_external/skipped_tests_numpy.tbl:2884]: > You can fallback on numpy for case when dtype is object.

Sorry, my fault. It is bad proposal, just leave it as is.
No need doing fallback on `numpy` for dparray own methods.

File: tests/skipped_tests.tbl [hunk 10]
```diff
@@ -154,7 +168,6 @@ tests/test_linalg.py::test_svd[(5,3)-complex128]
 tests/test_random.py::TestPermutationsTestShuffle::test_shuffle1[ lambda x: x]
 tests/test_random.py::TestPermutationsTestShuffle::test_shuffle1[ lambda x: dpnp.asarray(x).astype(dpnp.int8)]
 tests/test_random.py::TestPermutationsTestShuffle::test_shuffle1[ lambda x: dpnp.asarray(x).astype(dpnp.complex64)]
-tests/test_random.py::TestPermutationsTestShuffle::test_shuffle1[ lambda x: dpnp.asarray(x).astype(object)]
```
Comment (samir-nasibli) [inline tests/skipped_tests.tbl:157]: :

File: tests/test_dparray.py (hunk 6 above)
Comment (samir-nasibli) [reply tests/test_dparray.py:12]: outdated.

**Review by samir-nasibli — APPROVED** [review]
Looks pretty good. Thanks!

**Comment by samir-nasibli** [comment]
Issues with `complex` type can be solved on other PR.

File: dpnp/dpnp_algo/dpnp_algo.pyx [hunk 11]
```diff
@@ -125,10 +126,16 @@ cpdef dparray dpnp_array(obj, dtype=None):
 
 
 cpdef dparray dpnp_astype(dparray array1, dtype_target):
-    cdef dparray result = dparray(array1.shape, dtype=dtype_target)
+    cdef DPNPFuncType param1_type = dpnp_dtype_to_DPNPFuncType(array1.dtype)
+    cdef DPNPFuncType param2_type = dpnp_dtype_to_DPNPFuncType(dtype_target)
```
Comment (shssf) [inline dpnp/dpnp_algo/dpnp_algo.pyx:130]: `dtype_target` might be diffrent. It can contain some object with no `dtype`. I think this is the cause of the error you saw

**Review by shssf — APPROVED** [review]
